test/execute/index.js

```diff
@@ -0,0 +1,151 @@
+const fs = require('fs');
+const path = require('path');
+
+const test = require('tap').test;
+
+const log = require('../../src/util/log');
+const makeTestStorage = require('../fixtures/make-test-storage');
+const readFileToBuffer = require('../fixtures/readProjectFile').readFileToBuffer;
+const VirtualMachine = require('../../src/index');
+
+/**
+ * @fileoverview Transform each sb2 in this directory into a test.
+ *
+ * Test execution of a group of scratch blocks by SAYing if a test did "pass",
+ * or did "fail". Four keywords can be set at the beginning of a SAY messaage
+ * to indicate a test primitive.
+ *
+ * - "pass MESSAGE" will t.pass(MESSAGE).
+ * - "fail MESSAGE" will t.fail(MESSAGE).
+ * - "plan NUMBER_OF_TESTS" will t.plan(Number(NUMBER_OF_TESTS)).
+ * - "end" will t.end().
+ *
+ * A good strategy to follow is to SAY "plan NUMBER_OF_TESTS" first. Then
+ * "pass" and "fail" depending on expected scratch results in conditions, event
+ * scripts, or what is best for testing the target block or group of blocks.
+ * When its done you must SAY "end" so the test and tap know that the end has
+ * been reached.
+ */
+fs.readdirSync(__dirname)
+    .filter(uri => uri.endsWith('.sb2'))
+    .forEach(uri => {
+        test(uri, t => {
+            // Disable logging during this test.
+            log.suggest.deny('vm', 'error');
+            t.tearDown(() => log.suggest.clear());
+
+            const project = readFileToBuffer(path.resolve(__dirname, uri));
+
+            const vm = new VirtualMachine();
+            vm.attachStorage(makeTestStorage());
+
+            // Delay recording the early vm error with tap until plan has a
+            // chance to be set.
+            let vmError;
+            const testVmError = () => {
+                t.doesNotThrow(() => {
+                    if (vmError) {
+                        throw vmError;
+                    }
+                });
+            };
+
+            // Listen to SAY events and match the beginning of SAY text's to
+            // basic tap methods.
+            let didPlan;
+            let didEnd;
+            vm.runtime.on('SAY', (target, type, text) => {
+                if (text.toLowerCase().startsWith('pass')) {
```

Comment:
random thoughts, but after I'm done writing this, I don't think it's any more/less readable -- I kind of hate the repeat of toLowerCase for every one of the ifs here.

```js
const sayToTapReportMap = {
  pass: reason => t.pass(reason),
  fail: reason => t.fail(reason),
  plan: reason => {
    didPlan = true;
    t.plan(Number(reason) + 1);
  },
  end: () => {
    didEnd = true;
    if (!didPlan) {
      testVmError();
    }
    t.end();
  }
};

const unknownSay = reason => t.comment(reason);

(target, type, rawText) => {
  const text = rawText.toLowerCase();
  for (const [prefix, handler] of Object.entries(sayToTapReportMap)) {
    if (text.startsWith(prefix)) {
      return handler(rawText.substring(prefix.length));
    }
    unknownSay(rawText);
  }
  
}